File: packages/policy-api/policies/naive.lua
-- Naive baseline: how Surviving Mars: Relaunched trains appear to decide what to
-- carry, based on observed behaviour and not yet verified against the game.
--
-- At every stop, each resource the station stores is moved towards the average
-- stock of that resource across all stations on the line that store it. Nothing
-- else is considered: not how fast stations consume or produce, not what other
-- trains carry, and not which station needs the resource most.

local policy = {}

local function stores(station, resource)
  for _, id in ipairs(station.resources) do
    if id == resource then
      return true
    end
  end
  return false
end

function policy.on_stop(ctx)
  local station = ctx.station
  for _, resource in ipairs(station.resources) do
    local total, count = 0, 0
    for _, other in ipairs(ctx.line.stations) do
      if stores(other, resource) then
        total = total + other.stock[resource]
        count = count + 1
      end
    end
    local target = math.floor(total / count)
    local difference = station.stock[resource] - target
    if difference > 0 then
      ctx.load(resource, difference)
    elseif difference < 0 then
      ctx.unload(resource, -difference)
    end
  end
end

return policy

File: packages/engine/src/reference/naive.ts
import { emptyOutcome, type Policy, type PolicyOutcome } from "../policy.ts";

/**
 * Reference implementation of the naive baseline, used to cross-check
 * `naive.lua`: move every resource at the station towards the floor of the mean
 * stock of that resource across the stations that store it.
 */
export function naiveReferencePolicy(): Policy {
  return {
    start: () => emptyOutcome(),
    stop(snapshot): PolicyOutcome {
      const outcome = emptyOutcome();
      const here = snapshot.here;
      for (const resource of here.resources) {
        let total = 0;
        let count = 0;
        for (const id of snapshot.station_order) {
          const station = snapshot.stations[id];
          if (!station?.resources.includes(resource)) continue;
          if (station.stock === undefined) {
            return {
              ...emptyOutcome(),
              error: {
                kind: "runtime",
                message: "reading another station's stock requires the line information level",
              },
            };
          }
          total += station.stock[resource] ?? 0;
          count += 1;
        }
        const mean = Math.floor(total / count);
        const delta = (here.stock?.[resource] ?? 0) - mean;
        if (delta > 0) outcome.actions.push({ type: "load", resource, amount: delta });
        else if (delta < 0) outcome.actions.push({ type: "unload", resource, amount: -delta });
      }
      return outcome;
    },
  };
}
